### ISSUE TITLE
Cannot see what changes required to support SubprocVecEnv

### ISSUE THREAD

**Issue by davidwynter** [hill-a/stable-baselines#1079]
I have a custom environment that works with a single thread. I have read issue 334 here and the documentation here https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.htm but it is still not clear what changes I need to make.

This is not the full code, it uses the same constructs as my code and produces the same error. I know this is a lack of my understanding not a bug. It is unclear how the spaces.Box should be constructed, I read the documentation and by the method I use I ensure that the shape matches the state shape a single dimension array.
```
import numpy as np
import pandas as pd
from gym.utils import seeding
import gym
from gym import spaces
from stable_baselines.common.env_checker import check_env


class CustomEnv(gym.Env):
    """A custom environment for OpenAI gym"""
    metadata = {'render.modes': ['human']}

    def __init__(self, df, day = 0):
        self.day = day
        self.df = df

        # action_space normalization and shape is 30
        self.action_space = spaces.Box(low=-1, high=1, shape=(30,))
        # load data from a pandas dataframe
        self.data = self.df.loc[self.day, :]
        # initalize state
        self.state = np.array(np.random.randn(361), dtype=np.float32)
        self.observation_space = spaces.Box(low=0, high=np.inf, shape=(len(self.state),))
        self.terminal = False
        # initialize reward
        self.reward = 0
        self.cost = 0
        self.rewards = []
        # self.reset()
        self._seed()

    def step(self, actions):
        # print(self.day)
        self.terminal = self.day >= len(self.df.index.unique()) - 1
        # print(actions)

        if self.terminal:

            df_rewards = pd.DataFrame(self.rewards_memory)

            return self.state, self.reward, self.terminal, {}

        else:
            self.day += 1
            self.data = self.df.loc[self.day, :]
            # load next state, dummy
            self.state = np.array(np.random.randn(361), dtype=np.float32)

            self.reward = 100
            self.rewards_memory.append(self.reward)

        return self.state, self.reward, self.terminal, {}

    def reset(self):
        self.day = 0
        self.data = self.df.loc[self.day, :]
        self.cost = 0
        self.trades = 0
        self.terminal = False
        self.rewards_memory = []
        # initiate state
        self.state = np.array(np.random.randn(361), dtype=np.float32)
        # iteration += 1
        return self.state

    def render(self, mode='human'):
        return self.state

    def _seed(self, seed=None):
        self.np_random, seed = seeding.np_random(seed)
        return [seed]

train_df = pd.DataFrame(np.random.randint(0,100,size=(100, 361)))
train = CustomEnv(train_df)
check_env(train)
```

When I run with the 16 cpus I get the error below. Not sure it is related to the above but one thing at a time.

```
ValueError: could not broadcast input array from shape (16,1,361) into shape (16,361)
```

**System Info**
 Installed  using pip
 NVidia 1060 6GB
 Python version 3.7.9
 Tensorflow version 1.14.0 * 

Is there an example anywhere that shows a fully worked multiple process usage?

**Comment by araffin** [comment]
Hello,
please fill in the issue template completely, notably by checking your env using the env checker ;) (cf issue template and documentation)
For a custom env, you need at least to give the observation and action space (see [template](https://github.com/DLR-RM/stable-baselines3/blob/master/.github/ISSUE_TEMPLATE/custom_env.md) in our SB3 repo).

**Comment by davidwynter** [comment]
Solved this. 

Adding `check_env` fails because it detects either `DummyVecEnv` or `SubprocVecEnv` depending where you put the check_env() function in the code. 

The documentation should tell you that when you switch to use `SubprocVecEnv` you need to add a dimension to your state. Maybe I missed it? For me it was adding: 
`    out = np.expand_dims(out, axis=1)`
at the end of my method that sets the state.

**Comment by davidwynter** [comment]
I added the working example of my issue above. It is simple enough I think, but I just don't get what I need to do in the constructor for spaces.Box

**Comment by araffin** [comment]
> but I just don't get what I need to do in the constructor for spaces.Box

Please read openai Gym code for that ;) (this is not related to SB2)
https://github.com/openai/gym/blob/master/gym/spaces/box.py

>I added the working example of my issue above. 

Please follow completely the template in https://github.com/DLR-RM/stable-baselines3/blob/master/.github/ISSUE_TEMPLATE/custom_env.md
The current example cannot be run...
(step function is missing, output of check_env is missing and you should keep the `self.observation_space.sample()`)

**Comment by davidwynter** [comment]
I traced into the code in `check_env`. The message it produces is not really detailed enough:

```
"The observation returned by the `{}()` "
                                             "method does not match the given observation space".format(method_name)
```

The check is to actually see if the observation values are within the bounds of the Box, high and low values. Not sure why it does not say that is the issue. I had changed the calculation of the observation values and had not realised it was now producing values less than the low threshold. I learn't something.
